Formalization of Succinctness

ibbem opened the pull request:
This is the formalization of succinctness as advertised in my master thesis "On the Succinctness of
Languages for Static Variability".

Note that there are merge conflicts between `thesis_bm` and `main`. Hence, this branch already includes the merge to `main` and should be merged using a fast-forward merge.

ibbem committed:
Initial syntactic expressiveness experiments

This is the first definition has come to my mind. It works at least for
ADT < CCC intuitively.

ibbem committed:
Factor out the syntactic expressiveness definitions

ibbem committed:
Proof the expressiveness theorem for CCC instead of 2CC

I expect `2CC < CCC` so `ADT < CCC` should follow using transitivity.

ibbem committed:
Require ≤Size for <Size

As ≤Size is not total, <Size is not transitive without requiring ≤Size.
Moreover, ≱Size is not antisymmetric. Note that ≱Size is ¬ ≤Size with
the arguments flipped and the negation moved inside.

ibbem committed:
Prove that =Size, ≤Size and <Size behave as expected

ibbem committed:
Rename temporary files to have proper names

ibbem committed:
Move the size definitions into a separate file

ibbem committed:
Prove 2CC ≤ ADT to conclude 2CC < ADT

ibbem committed:
Prove that L₁ ≤Size L₂ and L₂ ≱Size L₁ cannot both be true

ibbem committed:
Use more representative variable names

ibbem committed:
Use shorter and more explicit notation for proofs

ibbem committed:
Prove CCC ≤ NCC

ibbem committed:
Prove 2CC < CCC

ibbem committed:
Quantify ≤Size over the artifact type

This makes it easier to use because the artifact type doesn't need to be
applied when invoking `≤Size`. Furthermore, this enables proofs of
`≱Size` to fix a single artifact type, for example the natural numbers,
and automatically have the inhabitants it needs.

The order between the quantifier over `n` and `A` doesn't have a big
impact. On the one hand, the chosen order allows `≱Size` to use
different artifact types for each `n`. However, it doesn't change the
relation inhabitants if they are swapped because there exists a type
with enough elements (i.e., union of all `A` ranging all `n`s) that can
be fixed and then only a subset of the artifacts can be used for a
specific `n`. On the other hand, `≤Size` is a `Set` and, thus, can't be
inspected if the order is changed. This specific order is chosen purely
as it's more convenient for pattern matching (e.g., one less `with`
clause in case of `≤Size`).

ibbem committed:
Remove unused imports

ibbem committed:
Cleanup the 2CC<ADT file

ibbem committed:
Make language arguments implicit

Previously, these where not inferred correctly, but now it works™.

ibbem committed:
Prove 2CC=2CC, NCC=2CC (for N=2) and conclude 2CC=CCC

ibbem committed:
Prove FST ≱ 2CC

ibbem committed:
Investigate the relationship between ≤Size and compilers

ibbem committed:
Prove our FST formalization misses a base feature

ibbem committed:
Use more specialized versions of ∷-injective

ibbem committed:
Prove OC ≱ 2CC

ibbem committed:
Use the artifact example definitions where possible

This reduces duplication and allows refactoring of 𝔸.

ibbem committed:
Redefine 𝔸 as a record

This allows to easily add more fields.

ibbem committed:
Implement atom sizes

ibbem committed:
Take advantage of artifact sizes

ibbem committed:
Rename `e₁` to `fst` and `e₂` to `2cc`

ibbem committed:
Factor out 2CC.reflectsVariantSize

ibbem committed:
Reuse List.replicate in OC≱2CC

ibbem committed:
Make OC≱2CC easier to read

ibbem committed:
Write 2CC instead of 2CC.2CC in FST≱2CC

ibbem committed:
Inline big-artifact in FST≱2CC

ibbem committed:
Use the same approach as OC≱2CC in FST≱2CC

Using subsets and uniqueness for partitioning is a little bit more
complicated but much shorter and more general. In the future,
`partition` will be factored into a separate module.

ibbem committed:
Simplify FST evaluation in FST≱2CC

ibbem committed:
Apply η-conversion in FST≱2CC

ibbem committed:
FST≱2CC: Only define variants that are needed

ibbem committed:
Make FST≱2CC more readable

ibbem committed:
FST≱2CC: Simplify `select-applyUpTo-feature`

This proof is much easier to digest as it mostly relies on reusable
theorems.

ibbem committed:
Factor out findings about fixed artifact lengths of 2CC

ibbem committed:
Move size>0 lemmas next to the size definitions

ibbem committed:
Factor out the inflation constant in OC≱2CC

ibbem committed:
Rename some things

ibbem committed:
Create size definitions using reflection

Note that currently indirect recursion is handled by special casing
`List` and `List⁺`. This will hopefully get fixed at some point.

ibbem committed:
Create a dead option elimination transformation

ibbem committed:
Rename SyntacticExpressiveness to Succinctness

ibbem committed:
Refactor the succinctness module structure

ibbem committed:
Add the designed definition

ibbem committed:
Inline size comparison

ibbem committed:
Use `SizedLang` in the designed succinctness relation

ibbem committed:
Split the relation definitions into separate files

ibbem committed:
Use the new succinctness definition

The designed succinctness definition includes a translatable constraint
that the old definition was missing. This gets rid of the unfortunate
`¬Compiler→¬≤` and `¬Compiler→≤` properties.

A drawback of this new definition is that it breaks transitivity.
Consider some languages L1 and L3 that are complete and a language L2
that is incomplete. There is an expression e in L1 that cannot be
translated to L2. If we have L1 <= L2 and L2 <= L3 we cannot conclude L1
<= L3 because we know nothing about the size of e translated to L3
because we just proved that there exists no translation to L2.

Note that the order of `∀ (A : 𝔸)` and `Σ[ m ∈ ℕ ]` was changed. Due to
parametricity (type parameters cannot be inspected) this does not change
the actual semantics of the definitions. However, it does simplify the
proofs by being friendlier to pattern matching and `with` clauses
avoiding additional helper functions in many cases.

ibbem committed:
Proof that there exists a diagonalization of ℕ

ibbem committed:
Prove that ≤ and ≰ are opposites

ibbem committed:
Prove ≤ₛ-weakening

ibbem committed:
Prove more general succinctness transitivity properties

ibbem committed:
Define an OC with a propositional selection language

ibbem committed:
Prove that OC is a subset of PropOC

ibbem committed:
Prove that PropOC is at least as succinct as OC

ibbem committed:
Remove annoying module prefixes in 2CC<ADT

ibbem committed:
Add missing size definitions

ibbem committed:
Generalize the coarseness of 2CC < ADT

ibbem committed:
Add some more transitivity theorems

ibbem committed:
Simplify the FixedArtifactLength lemma

ibbem committed:
Replace ≱ in names with ≰

This makes the names consistent with the symbol that is used now.

ibbem committed:
Generalize the dimension of 2CC ≰ₛ FST

ibbem committed:
Generalize the dimension of 2CC ≰ₛ OC

ibbem committed:
Generalize the dimension of 2CC <ₛ ADT

ibbem committed:
Prove VariantList ≤ₛ ADT

ibbem committed:
Prove NADT ≤ₛ VariantList

ibbem committed:
Prove ADT ≤ₛ NADT

ibbem committed:
Prove NADT ≤ₛ ADT

ibbem committed:
Generalize the coarseness of the relationship between ≰ and ≤

ibbem committed:
Prove independence of size definition

ibbem committed:
Remove a lemma that is available in the standard library

pmbittner commented on src/Vatras/Framework/Definitions.agda:
```diff
@@ -20,12 +23,13 @@ the core definitions because it is quite reasonable.
-atoms : 𝔸 → Set
-atoms = proj₁
+record 𝔸 : Set₁ where
+  no-eta-equality
```
Maybe you could document here why eta equality must be forbidden?

ibbem replied:
I'm actually unsure why exactly this is needed. However, when changing `𝔸` from `data` to a `record`, there was a proof that failed suddenly. As far as I understand eta-equality, it should only increase expressiveness. Hence, this could potentially be a limitation of bug of Agda. However, we do not need eta-equality so I decided to just disable it (like it was before, with `data`).

pmbittner replied:
I see :) Thank you for explaining it.

pmbittner commented on src/Vatras/Framework/Definitions.agda:
```diff
@@ -63,14 +67,23 @@ and hence expressions are parameterized in the type of this atomic data.
-  open import Data.Nat using (ℕ; _≟_)
+NAT : 𝔸
+NAT = record
+  { atoms = ℕ × ℕ
```
Could you document why there are two natural numbers? What is each number supposed to mean?

pmbittner commented on src/Vatras/Lang/2CC/FixedArtifactLength.agda:
```diff
@@ -0,0 +1,118 @@
+  → (vs : List (Rose ∞ A))
+  → AllPairs (_≉_) vs
+  → All (_∈ ⟦ e ⟧) vs
+  → List.sum (List.map sizeRose vs) ≤ size2CC e
```
This seems rather counter-intuitive. A plain enumeration of trees is smaller absolute size than a single 2CC expression which potentially reduces duplication? Am I misreading this theorem? Could you add some documentation?

ibbem replied:
The key requirement for this lemma is `AllPairs (_≉_) vs` which expresses that the number of top-level children of the variants `vs` are pairwise distinct. Thus choice calculus (without `let` or `macro`) cannot share any sub-expressions.

pmbittner replied:
Ah yes this makes total sense. Could you add the comment you made here as documentation? :)

pmbittner commented on src/Vatras/Lang/FST/NoCoreFeature.agda:
```diff
@@ -0,0 +1,51 @@
+does-not-describe-variant (e , variant⊆e , e⊆variant) | zero , e≡variant | a , e≡empty | ()
+
+FST-is-incomplete : Incomplete (Rose ∞) (FST.FSTL F)
+FST-is-incomplete complete = does-not-describe-variant (Prod.map₂ (≅-sym) (complete variantGenerator))
```
Maybe you could add a sentence of documentation to this module what the main insight is here. Is this the first (direct) proof of FST incompleteness?

pmbittner commented on src/Vatras/Translation/Lang/OC/DeadElim.agda:
```diff
@@ -0,0 +1,168 @@
+open import Vatras.Lang.OC F using (OC; _-<_>-; _❲_❳; Configuration; ⟦_⟧ₒ)
+open import Vatras.Lang.OC.Util using (all-oc)
+
+data RestrictOptions {A : 𝔸} : {i : Size} → List F → OC i A → Set₁ where
```
What does this do? Restrict options to not be in a given environment (i.e., list of options)?

pmbittner commented on src/Vatras/Translation/Lang/OC/DeadElim.agda:
```diff
@@ -0,0 +1,168 @@
+elimDead-preserves : {i : Size} → {A : 𝔸} → (e : OC i A) → ⟦ elimDead e ⟧ₒ ≅[ id ][ id ] ⟦ e ⟧ₒ
+elimDead-preserves e = ≗→≅[] (λ c → elimDead-preserves' [] e c [])
+
+-- TODO WFOC
```
Is this TODO still up to date? It can stay here if it is.

ibbem replied:
Yes, there might be use cases for adding a wrapper for undead `WFOC`.

pmbittner replied:
Ok then let's leave it here.

pmbittner commented on src/Vatras/Lang/OC/DeadElim.agda:
```diff
@@ -0,0 +1,168 @@
+
+elimDead' : {i : Size} → {A : 𝔸} → (env : List F) → OC i A → Σ (OC ∞ A) (RestrictOptions env)
+elimDead' env (a -< cs >-) = a -< List.map proj₁ (List.map (elimDead' env) cs) >- , (env -< All.fromList (List.map (elimDead' env) cs) >-)
+elimDead' env (a ❲ c ❳) with a ∈? env
```
It would be more consistent with the rest of the library to call variables of options `O` instead of `a` but this isn't too important.

pmbittner commented on src/Vatras/Translation/Lang/OC/DeadElim.agda:
```diff
@@ -0,0 +1,168 @@
+undead→≢ : {i : Size} → {A : 𝔸} → {f₁ f₂ : F} → {e : OC i A} → Undead (f₁ ❲ f₂ ❲ e ❳ ❳) → f₁ ≢ f₂
+undead→≢ (undead (f₁∉env ❲ f₂∉env ❲ undead-e ❳ ❳)) f₁≡f₂ = f₂∉env (here (Eq.sym f₁≡f₂))
+
+changeConfig : F → Bool → Configuration → Configuration
```
What is this supposed to do? Can you recall the idea of this part of the preservation proof?

ibbem replied:
This is not relevant for the preservation proof. Instead I used this lemma while trying to prove some succinctness property (essentially something like a translation to a minimal equivalent expression) about option calculus. However, I'm not sure if that approach is actually correct and it didn't have the time to work more on that proof.
Currently, this whole module is unused in Vatras. However, I still included it because I already did the work and it might become useful for somebody in the future.

pmbittner replied:
Ok fine with me. Could you add this explanation as module documentation?

pmbittner commented on src/Vatras/Lang/PropOC.agda:
```diff
@@ -0,0 +1,37 @@
+open import Vatras.Framework.Definitions using (𝔽; ℂ; 𝔼)
+
+module Vatras.Lang.PropOC (F : 𝔽) where
```
Maybe a short note that this module implements option calculus with propositional formulas as option values?

pmbittner commented on src/Vatras/Succinctness/Relations/2CC<ADT.agda:
```diff
@@ -0,0 +1,576 @@
+  where
+  open Eq.≡-Reasoning
+
+todo5 :
```
Should we keep these `todo...` names? Looks like you wanted to give hem a name?

ibbem replied:
Oh, ... Embarrassing 😅
I'll change these names.

pmbittner replied:
Nothing to be ashamed about. :)

pmbittner commented on src/Vatras/Succinctness/Relations/2CC=2CC.agda:
```diff
@@ -0,0 +1,153 @@
+module Vatras.Succinctness.Relations.2CC=2CC where
```
A short note what this module is about? The name is rather irritating :D

pmbittner commented on src/Vatras/Util/Big-O.agda:
```diff
@@ -0,0 +1,68 @@
+open import Data.Product using (_,_; Σ-syntax)
+open import Function using (id)
+import Relation.Binary.PropositionalEquality as Eq
+open import Relation.Unary using (_∈_)
```
I believe this module is more than just utility isn't it? Would it make sense to move it to the Succinctness module? What do you think? In any case, we can leave it here of course to save some work.

ibbem replied:
My rational for putting it into the `Util` package is that, from the point of view of Vatras, plain big-O notation is just an existing mathematical framework that we make use of (i.e., it doesn't really belong to Vatras per-se). Just because, right now, Vatras has only one use case of big-O notation but it could, in theory, have a second use case outside of Succinctness. Then it wouldn't make sense to have big-O notation in the `Succinctness` module.

pmbittner replied:
This line of reasoning makes sense to me.

pmbittner commented on src/Vatras/Util/Nat/Diagonalization.agda:
```diff
@@ -0,0 +1,214 @@
+{-# OPTIONS --allow-unsolved-metas #-}
```
unsolved-metas still necessary?

pmbittner commented on src/Vatras/Util/Nat/Diagonalization.agda:
```diff
@@ -0,0 +1,214 @@
+
+import Vatras.Util.List as List
+
+diagonalization : ℕ × ℕ → ℕ
```
Can you add some documentation of what this diagonalization module is doing?

pmbittner reviewed (approved):
Hi @ibbem 

thank you for the PR. As always, it would be great to have some documentation here and there. I leave it up to you to decide how much time you want to invest. I have some comments below but would be willing to merge in any case.

ibbem committed:
Remove a superfluous `--allow-unsolved-metas`

ibbem committed:
Rename some variables and lemmas in 2CC<ADT

ibbem committed:
Move OC dead elimination to `Translation`

ibbem committed:
Use typical option names for OC dead elimination

ibbem committed:
Add some useful documentation

ibbem committed:
Add my thesis to the README

ibbem commented:
I added all the requested [documentation](https://github.com/VariantSync/Vatras/commit/083283187d0506bbb62359e40c8a2d9a0170c7d9) and added my thesis in the [README](https://github.com/VariantSync/Vatras/commit/f324d067bdf8ad73fb75790018f15b0d1fafff0c).

pmbittner commented on src/Vatras/Lang/2CC/FixedArtifactLength.agda:
```diff
@@ -1,3 +1,8 @@
+{-|
+This module contains some lemmas that make the fact,
+that artifacts of choice calculus expressions have a fixed number of children
+usable.
```
What does it mean to "have a child usable"?

ibbem replied:
The lemma is more usable. I forgot a comma at the end of the line there.
I rewrote that sentence to be more understandable.

pmbittner commented on src/Vatras/Lang/FST/NoBaseArtifacts.agda:
```diff
@@ -1,3 +1,10 @@
+{-|
+Proof that feature structure trees do not contain a base artifact.
```
Feature structure trees do have a base artifact? There is a shared root onto which all trees are imposed, isn't it? Or did I get something wrong?

pmbittner replied:
My understanding here is: The `SPL` type, which encodes expressions of this variability language, has an artifact and that artifact is the root of every produced variant. Isn't that a base artifact?

ibbem replied:
Right. There is always an artifact. Effectively, I mean that there is no core feature.
I renamed the module accordingly.

pmbittner commented on src/Vatras/Translation/Lang/OC/DeadElim.agda:
```diff
@@ -104,6 +142,9 @@ changeConfig-∉ f f∈env b c (f' ❲ e ❳) undead-e with f ≟ f'
 changeConfig-∉ f f∈env b c (f' ❲ e ❳) (f'∉env ❲ undead-e ❳) | no f≢f' = Eq.cong (λ e → if c f' then e else nothing) (changeConfig-∉ f (there f∈env) b c e undead-e)
 
+{-|
+An option whoose dimensions is configured to `true` is semantically equivalent to its child.
```
typo `whoose`

pmbittner reviewed (approved):
Thank you a lot for the documentation! I have a few last questions on the new documentation, and then we can merge. :)

pmbittner commented on README.md:
```diff
@@ -130,6 +130,14 @@ Details on the features implemented in Vatras, including tutorials for integrati
+
+> Benjamin Moosherr. _On the Succinctness of Languages for Static Variability_. Master Thesis, University of Ulm, November 2025. Reviewed by Matthias Tichy and Thomas Thüm. Supervised by Paul Maximilian Bittner.
+
+This Master thesis extends Vatras by a succinctness relation. Succinctness expresses how the size of expressions must change when translated using a variability language compiler. Hence, succinctness allows us to differentiate between variability languages which have equal semantic expressiveness but whose translation results in combinatorial explosion.
```
Great! I think it must be spelled `master thesis` in lower case here. 🤔

ibbem committed:
Rename NoBaseArtifacts into NoCoreFeature

The name was misleading as every variant has a root artifact (although
it doesn't have any children). The new name also incorporates existing
software product line terminology.

ibbem committed:
Merge branch 'develop' into thesis_bm
